### PR TITLE
Deleted students trying to join: handle the NPE #5931

### LINKED ISSUE #5931: Deleted students trying to join: handle the NPE
If a student was deleted from a course, when the student tries to join using the join link, we throw a NullPointerException.

### PR DESCRIPTION
**Pull request by Neurrone**
Fixes #5931

### CODE REVIEW

**Comment by Neurrone** [comment]
WIP. The NPE, no longer happens, but redirecting to student homepage is a placeholder as it doesn't show the desired status message.

Is there an existing page that shows status messages without requiring authentication?

**Comment by JosephineKwa** [comment]
Do also post screenshots where applicable :)

File: src/main/java/teammates/common/util/Const.java [hunk 1]
```diff
@@ -1396,7 +1396,8 @@ private Const() {
         public static final String STUDENT_UPDATE_PROFILE_NATIONALITY =
                 "Meanwhile, you can provide your nationality "
                 + "<a href=\"" + Const.ActionURIs.STUDENT_PROFILE_PAGE + "\">here</a>.";
-        
+        public static final String DELETED_STUDENT_ATTEMPTING_TO_JOIN =
+                "You can no longer join the course, as you you have been unenrolled from the course by an instructor.";
```
Comment (Neurrone) [inline src/main/java/teammates/common/util/Const.java]: Oh yeah, i forgot that when I was testing that, i wasn't logged in as a student, so I thought I did something wrong. Works as expected now.

**Comment by Neurrone** [comment]
Changes made.

File: src/main/java/teammates/common/util/Const.java [hunk 2]
```diff
@@ -1403,6 +1403,9 @@ private Const() {
         public static final String STUDENT_COURSE_JOIN_SUCCESSFUL = "You have been successfully added to the course %s.";
         
         /** Template String. Parameters:  Course ID */
+        public static final String DELETED_STUDENT_ATTEMPTING_TO_JOIN =
+                "Unable to join course %s as you have been unenrolled by an instructor. Please contact your "
+                        + "course instructor for assistance.";
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/util/Const.java]: The indentation here should be 8 spaces, i.e. `+ "course ...` should be aligned with `"Unable to ...`

**Comment by Neurrone** [comment]
Changes made. Would this happening also require that the returned result has its isError flag set, since its using the warning colour?

**Comment by JosephineKwa** [comment]
Yes @Neurrone, do set the `isError`flag 👍

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java [hunk 3]
```diff
@@ -20,12 +22,22 @@
...
-        String nextUrl = getNextUrl();
         
         statusToAdmin = "Action Student Clicked Join Link"
                         + (account.googleId == null ? "<br>Email: " + account.email
                                                     : "<br>Google ID: " + account.googleId + "<br>Key: " + regkey);
         
+        if (student == null) {
+            statusToAdmin += "<br>Student course join failed due to being deleted by instructor.";
+            String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
+            Assumption.assertNotNull(courseId);
```
Comment (JosephineKwa) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: Perhaps `assertPostParamNotNull` would be more apt here

Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: Let this one be for now. @Neurrone if you're interested, #5532 #2972 are for you.

**Comment by Neurrone** [comment]
Changes made.

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java (hunk 3 above)
Comment (Neurrone) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: Thanks, will take a look.

File: src/main/java/teammates/common/util/Const.java [hunk 4]
```diff
@@ -1403,6 +1403,9 @@ private Const() {
         public static final String STUDENT_COURSE_JOIN_SUCCESSFUL = "You have been successfully added to the course %s.";
         
         /** Template String. Parameters:  Course ID */
+        public static final String DELETED_STUDENT_ATTEMPTING_TO_JOIN =
+                "Unable to join course %s as you have been unenrolled by an instructor. Please contact your "
+                + "course instructor for assistance.";
```
Comment (damithc) [inline src/main/java/teammates/common/util/Const.java]: We may need to rephrase the message to avoid situation where a student email the instructor "How come you unenrolled me?" and the instructor replies "I did no such thing". In rare occasions, this situation can be caused by other issues such as eventual consistency, so it is better to phrase the message is a non-committal way.

How about this?

`Unable to join course %s as you are currently not in the student list of that course. Please contact your course instructor for assistance.`

**Comment by Neurrone** [comment]
@damithc reworded.

File: src/test/java/teammates/test/cases/action/StudentCourseJoinActionTest.java [hunk 5]
```diff
@@ -106,7 +106,25 @@ public void testExecuteAndPostProcess() throws Exception {
         
         // delete the new student
         studentsDb.deleteStudentWithoutDocument(newStudentData.course, newStudentData.email);
+        
+        ______TS("Unenrolled student joining course displays error");
```
Comment (JosephineKwa) [inline src/test/java/teammates/test/cases/action/StudentCourseJoinActionTest.java]: Do reword this as well, to `Non-existent student attempting to join course`

File: src/main/java/teammates/common/util/Const.java [hunk 6]
```diff
@@ -1403,6 +1403,9 @@ private Const() {
         public static final String STUDENT_COURSE_JOIN_SUCCESSFUL = "You have been successfully added to the course %s.";
         
         /** Template String. Parameters:  Course ID */
+        public static final String DELETED_STUDENT_ATTEMPTING_TO_JOIN =
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/util/Const.java]: Rename this constant as well, to `NON_EXISTENT_STUDENT_ATTEMPTING_TO_JOIN_COURSE`

File: src/main/java/teammates/ui/controller/StudentCourseJoinAction.java [hunk 7]
```diff
@@ -20,12 +22,23 @@
...
     public ActionResult execute() {
         Assumption.assertPostParamNotNull(Const.ParamsNames.REGKEY, regkey);
-        String nextUrl = getNextUrl();
         
         statusToAdmin = "Action Student Clicked Join Link"
                         + (account.googleId == null ? "<br>Email: " + account.email
                                                     : "<br>Google ID: " + account.googleId + "<br>Key: " + regkey);
         
+        if (student == null) {
+            statusToAdmin += "<br>Student course join failed due to being deleted by instructor.";
```
Comment (JosephineKwa) [inline src/main/java/teammates/ui/controller/StudentCourseJoinAction.java]: Reword this status as well, to `<br>Student course join failed as student does not exist.`

**Comment by Neurrone** [comment]
@JosephineKwa reworded.
